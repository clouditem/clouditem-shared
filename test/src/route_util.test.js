
const supertest = require('supertest'),
  // {unit_test} = require('../util.js'),
  sut = require('./../../src/route_util.js'),
  express = require('express'),
  send = str => (req, res) => res.status(200).send(str),
  mirror = fn => (req, res) => res.status(200).send(fn(req)),
  reject = str => () => Promise.reject(Error(str));

describe('routing', () => {

  test.each`
  input                                                           | req_method  | req_path          |  test_name
  ${{post: {'/path': send('done')}}}                              |  ${'post'}  | ${'/path'}        | ${'simple path'}
  ${{get: {'/:param': mirror(re => re.params.param)}}}            |  ${'get'}   | ${'/abrakadabra'} | ${'parameter mirror'}
  ${{get: {'/:param': mirror(re => re.params.param)}}}            |  ${'get'}   | ${'/'}            | ${'parameter mirror'}
  ${{get: {'/:param': send('/:param'), '/': send('root')}}}       |  ${'get'}   | ${'/'}            | ${'parameter root'}
  ${{get: {'/:param': send('/:param'), '/': send('root')}}}       |  ${'get'}   | ${'/abcdf'}       | ${'parameter root'}
  ${{get: {'/': send('root'), '/:param': send('/:param')}}}       |  ${'get'}   | ${'/abcdf'}       | ${'root parameter'}
  ${{get: {'/': send('root'), '/:param': send('/:param')}}}       |  ${'get'}   | ${'/'}            | ${'root parameter'}
  ${{get: {'/path': send('path'), '/:param': send('/:param')}}}   |  ${'get'}   | ${'/path'}         | ${'path parameter'}
  ${{get: {'/:param': send('/:param'), '/path/': send('path')}}}  |  ${'get'}   | ${'/path'}         | ${'parameter path'}
  ${{get: {'/sub/*': send('sub'), '/:param': send('/:param')}}}   |  ${'get'}   | ${'/st/sub'}      | ${'subpath parameter'}
  ${{get: {'/': reject('test_problem')}}}                         |  ${'get'}   | ${'/'}            | ${'rejection'}
  ${{error_handlers: {Error: (res, err) => res.send(`from handler: ${err}`)}, get: {'/': reject('test_problem')}}}  |  ${'get'}   | ${'/'} | ${'rejection with error handler'}
  `('$test_name: [$req_method $req_path]', ({input, req_method, req_path}) => {

  const app = express();
  app.use(sut.create_router_according_to_mapping(input));
  return Promise.resolve(supertest(app)[req_method](req_path))
    .then(({text, statusCode: code}) => expect({text, code}).toMatchSnapshot());

});

});

describe('middleware', () => {

  test.each`
  input                                                                     | req_body     |  test_name
  ${{json: undefined, promise: (req, res) => res.send(req.body)}}           |  ${'{}'}     | ${'json'}
  ${{json: '10mb', promise: (req, res) => res.send(req.body)}}              |  ${'{}'}     | ${'json with limit'}
  ${{promise: (req, res) => res.send(req.body), json: '10mb'}}              |  ${'{}'}     | ${'order reverse fail'}
  ${{ensure_logged_in: true, promise: (req, res) => res.send(req.body)}}    |  ${'{}'}     | ${'unauthorized redirects to login'}
  `('$test_name: [$req_body]', ({input, req_body}) => {

  const app = express();
  app.use(sut.create_router_according_to_mapping({post: {'/target': input}}));
  return Promise.resolve(supertest(app).post('/target')
    .send(req_body)).then(({text, statusCode: code}) => expect({text, code}).toMatchSnapshot());

});

  test.each`
  input                                                                         | req_headers     |  test_name
  ${{cors: undefined, promise: (req, res) => res.send(req.body)}}               |  ${{}}          | ${'cors'}
  ${{cors: {origin: 'ioot.org'}, promise: (req, res) => res.send(req.body)}}    |  ${{}}          | ${'cors allow ioot.org'}
  ${{promise: (req, res) => res.send(req.body)}}                                |  ${{}}          | ${'blank'}
`('$test_name: [$req_headers]', ({input, req_headers}) => {

  const app = express();
  app.use(sut.create_router_according_to_mapping({post: {'/target': input}}));
  return Promise.resolve(supertest(app).post('/target')).then(({header}) => expect(header).toMatchSnapshot({date: expect.anything()}));

});

  test('cors at root_level', () => {

    const app = express();
    app.use(sut.create_router_according_to_mapping({cors: undefined, post: {'/target': (req, res) => res.send('done')}}));
    return Promise.resolve(supertest(app).post('/target')).then(({header}) => expect(header).toMatchSnapshot({date: expect.anything()}));

  });
  describe('parameters', () => {

    test.each`
  request_parameters                              | req_path                                    |    req_body                           |   test_name
  ${['path_param', 'body_param', 'query_param']}  | ${'/from_path?query_param=from query'}      |    ${{body_param: {from: 'body'}}}    | ${'all'}
  ${['path_param']}                               | ${'/from_path?path_param=from query'}       |    ${{path_param: {from: 'body'}}}    | ${'path is first'}
  ${['body_param']}                               | ${'/from_path'}                             |    ${{}}                              | ${'error throws error message'}
  ${['body_param']}                               | ${'/from_path?body_param=from query'}       |    ${{body_param: {from: 'body'}}}    | ${'body is second'}
  ${['query_param']}                              | ${'/from_path?query_param=from query'}      |    ${{body_param: {from: 'body'}}}    | ${'query is third'}
  `('$test_name: $request_parameters', ({request_parameters, req_path, req_body}) => {

  const app = express();
  app.use(sut.create_router_according_to_mapping({post: {'/:path_param': {json: undefined, request_parameters, promise: ({promised_params}, res) => res.send({promised_params})}}}));
  return Promise.resolve(supertest(app).post(req_path)
    .send(req_body))
    .then(({text}) => expect(text).toMatchSnapshot());

});

  });

});

test('wrong methods do not throw exception and do not show up in possible methods', () => {

  const app = express();
  app.use(sut.create_router_according_to_mapping({ignored: {'/': send('this_should_not_possibly_happen')}}));
  return Promise.resolve(supertest(app).get('/'))
    .then(({text, statusCode}) => expect({text, statusCode}).toMatchSnapshot({statusCode: 405, text: expect.not.stringContaining('ignored')}));

});
