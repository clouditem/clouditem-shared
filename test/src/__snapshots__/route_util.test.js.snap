// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`middleware "blank": [{}] 1`] = `
Object {
  "connection": "close",
  "content-length": "0",
  "date": Anything,
  "x-powered-by": "Express",
}
`;

exports[`middleware "cors allow ioot.org": [{}] 1`] = `
Object {
  "access-control-allow-origin": "ioot.org",
  "connection": "close",
  "content-length": "0",
  "date": Anything,
  "vary": "Origin",
  "x-powered-by": "Express",
}
`;

exports[`middleware "cors": [{}] 1`] = `
Object {
  "access-control-allow-origin": "*",
  "connection": "close",
  "content-length": "0",
  "date": Anything,
  "x-powered-by": "Express",
}
`;

exports[`middleware "json with limit": ["{}"] 1`] = `
Object {
  "code": 200,
  "text": "{}",
}
`;

exports[`middleware "json": ["{}"] 1`] = `
Object {
  "code": 200,
  "text": "{}",
}
`;

exports[`middleware "order reverse fail": ["{}"] 1`] = `
Object {
  "code": 200,
  "text": "",
}
`;

exports[`middleware "unauthorized redirects to login": ["{}"] 1`] = `
Object {
  "code": 302,
  "text": "Found. Redirecting to /login",
}
`;

exports[`middleware cors at root_level 1`] = `
Object {
  "access-control-allow-origin": "*",
  "connection": "close",
  "content-length": "4",
  "content-type": "text/html; charset=utf-8",
  "date": Anything,
  "etag": "W/\\"4-5f2c/g6AOREdVLWI53srsMrUHDo\\"",
  "x-powered-by": "Express",
}
`;

exports[`middleware parameters "all": ["path_param", "body_param", "query_param"] 1`] = `"{\\"promised_params\\":{\\"path_param\\":\\"from_path\\",\\"body_param\\":{\\"from\\":\\"body\\"},\\"query_param\\":\\"from query\\"}}"`;

exports[`middleware parameters "body is second": ["body_param"] 1`] = `"{\\"promised_params\\":{\\"body_param\\":{\\"from\\":\\"body\\"}}}"`;

exports[`middleware parameters "error throws error message": ["body_param"] 1`] = `"from handler: Error: Request does not contain value [body_param] in any of the following fields: [params,body,query]"`;

exports[`middleware parameters "path is first": ["path_param"] 1`] = `"{\\"promised_params\\":{\\"path_param\\":\\"from_path\\"}}"`;

exports[`middleware parameters "query is third": ["query_param"] 1`] = `"{\\"promised_params\\":{\\"query_param\\":\\"from query\\"}}"`;

exports[`routing "parameter mirror": ["get" "/"] 1`] = `
Object {
  "code": 404,
  "text": "Path [/] was not found, try [\\"/:param\\"] instead or use another method",
}
`;

exports[`routing "parameter mirror": ["get" "/abrakadabra"] 1`] = `
Object {
  "code": 200,
  "text": "abrakadabra",
}
`;

exports[`routing "parameter path": ["get" "/path"] 1`] = `
Object {
  "code": 200,
  "text": "/:param",
}
`;

exports[`routing "parameter root": ["get" "/"] 1`] = `
Object {
  "code": 200,
  "text": "root",
}
`;

exports[`routing "parameter root": ["get" "/abcdf"] 1`] = `
Object {
  "code": 200,
  "text": "/:param",
}
`;

exports[`routing "path parameter": ["get" "/path"] 1`] = `
Object {
  "code": 200,
  "text": "path",
}
`;

exports[`routing "rejection with error handler": ["get" "/"] 1`] = `
Object {
  "code": 200,
  "text": "from handler: Error: test_problem",
}
`;

exports[`routing "rejection": ["get" "/"] 1`] = `
Object {
  "code": 500,
  "text": "Internal Server Error
Error: test_problem",
}
`;

exports[`routing "root parameter": ["get" "/"] 1`] = `
Object {
  "code": 200,
  "text": "root",
}
`;

exports[`routing "root parameter": ["get" "/abcdf"] 1`] = `
Object {
  "code": 200,
  "text": "/:param",
}
`;

exports[`routing "simple path": ["post" "/path"] 1`] = `
Object {
  "code": 200,
  "text": "done",
}
`;

exports[`routing "subpath parameter": ["get" "/st/sub"] 1`] = `
Object {
  "code": 404,
  "text": "Path [/st/sub] was not found, try [\\"/sub/*\\",\\"/:param\\"] instead or use another method",
}
`;

exports[`wrong methods do not throw exception and do not show up in possible methods 1`] = `
Object {
  "statusCode": 405,
  "text": StringNotContaining "ignored",
}
`;
